### PR TITLE
Fixes and refactoring for LayerMap

### PR DESCRIPTION
**Pull request by hlinnaka**
I started working on https://github.com/zenithdb/zenith/issues/517 but bumped into other issues, and found some refactorings that I think makes the code nicer. This PR includes those things. It doesn't fix https://github.com/zenithdb/zenith/issues/517 yet, I'll continue working on that.

### CODE REVIEW

File: pageserver/src/layered_repository/layer_map.rs [hunk 1]
```diff
@@ -42,62 +41,24 @@ pub struct LayerMap {
     /// All in-memory layers, ordered by 'oldest_pending_lsn' of each layer.
```
Comment (lubennikovaav) [inline pageserver/src/layered_repository/layer_map.rs:41]: Should this comment mention `generation` too?

File: pageserver/src/layered_repository/layer_map.rs [hunk 2]
```diff
@@ -340,16 +236,114 @@ impl LayerMap {
...
+    }
+}
+
+/// Entry held in LayerMap::open_layers, with boilerplate comparison
+/// routines to implement a min-heap ordered by 'oldest_pending_lsn' and 'generation''
+///
+/// Each entry also carries a generation number. It can be used to distinguish
+/// entries with the same 'oldest_pending_lsn'.
+struct OpenLayerEntry {
+    pub oldest_pending_lsn: Lsn,
```
Comment (lubennikovaav) [inline pageserver/src/layered_repository/layer_map.rs:320]: Please, add a comment to `oldest_pending_lsn`

**Review by lubennikovaav — APPROVED** [review]

File: pageserver/src/layered_repository/layer_map.rs [hunk 3]
```diff
@@ -132,67 +77,53 @@ impl LayerMap {
...
-                open: Some(Arc::clone(&layer)),
-                historic: BTreeMap::new(),
-            };
-            self.segs.insert(tag, segentry);
-        }
+        segentry.insert_open(Arc::clone(&layer));
 
-        let opensegentry = OpenSegEntry {
+        // Also add it to the binary heap
+        let opensegentry = OpenLayerEntry {
```
Comment (patins) [inline pageserver/src/layered_repository/layer_map.rs:93]: `open_layer_entry`

File: pageserver/src/layered_repository/layer_map.rs [hunk 4]
```diff
@@ -132,67 +77,53 @@ impl LayerMap {
...
+        self.open_layers.push(opensegentry);
 
         NUM_INMEMORY_LAYERS.inc();
     }
 
     /// Remove the oldest in-memory layer
     pub fn pop_oldest_open(&mut self) {
-        let opensegentry = self.open_segs.pop().unwrap();
+        // Pop it from the binary heap
+        let opensegentry = self.open_layers.pop().unwrap();
```
Comment (patins) [inline pageserver/src/layered_repository/layer_map.rs:106]: `open_layer_entry`

File: pageserver/src/layered_repository/layer_map.rs [hunk 5]
```diff
@@ -269,38 +188,15 @@ impl LayerMap {
...
+            false
         }
-        trace!("no later layer found for {}, {}", seg, lsn);
-        false
     }
 
     /// Return the oldest in-memory layer, along with its generation number.
     pub fn peek_oldest_open(&self) -> Option<(Arc<InMemoryLayer>, u64)> {
-        if let Some(opensegentry) = self.open_segs.peek() {
+        if let Some(opensegentry) = self.open_layers.peek() {
```
Comment (patins) [inline pageserver/src/layered_repository/layer_map.rs:199]: `open_layer_entry`

File: pageserver/src/layered_repository/layer_map.rs [hunk 6]
```diff
@@ -340,16 +236,114 @@ impl LayerMap {
...
+        if let Some(open) = &self.open {
+            if open.get_start_lsn() <= lsn {
+                let x: Arc<dyn Layer> = Arc::clone(open) as _;
+                return Some(x);
+            }
+        }
+
+        if let Some((_start_lsn, layer)) = self
+            .historic
+            .range((Included(Lsn(0)), Included(lsn)))
```
Comment (patins) [inline pageserver/src/layered_repository/layer_map.rs:272]: Can we do `..=lsn`

File: pageserver/src/layered_repository/layer_map.rs [hunk 7]
```diff
@@ -340,16 +236,114 @@ impl LayerMap {
...
+        }
+    }
+
+    pub fn newer_image_layer_exists(&self, lsn: Lsn) -> bool {
+        // We only check on-disk layers, because
+        // in-memory layers are not durable
+
+        for (_newer_lsn, layer) in self
+            .historic
+            .range((Included(lsn), Included(Lsn(u64::MAX))))
```
Comment (patins) [inline pageserver/src/layered_repository/layer_map.rs:287]: `lsn..`

**Review by patins — APPROVED** [review]

File: pageserver/src/layered_repository/layer_map.rs (hunk 1 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/layer_map.rs]: added

File: pageserver/src/layered_repository/layer_map.rs [hunk 8]
```diff
@@ -42,62 +41,24 @@ pub struct LayerMap {
...
+            segs: HashMap::new(),
+            open_layers: BinaryHeap::new(),
+            current_generation: 0,
+        }
     }
 }
-impl Eq for OpenSegEntry {}
 
-impl LayerMap {
+impl<'a> LayerMap {
```
[the remark this answers is not in the document]
Comment (hlinnaka) [reply pageserver/src/layered_repository/layer_map.rs:60]: apparently not, removed

File: pageserver/src/layered_repository/layer_map.rs (hunk 3 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/layer_map.rs]: renamed to `oldest_entry`

File: pageserver/src/layered_repository/layer_map.rs (hunk 6 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/layer_map.rs]: I was a bit reluctant to do that, it looks very terse, and I'm used to the syntax. But after staring at it for a few minutes and getting used to it, I agree it's better that way. Changed.

File: pageserver/src/layered_repository/layer_map.rs (hunk 2 above)
Comment (hlinnaka) [reply pageserver/src/layered_repository/layer_map.rs]: There are some comments in `InMemoryLayer` for it. But I added a brief `// copy of layer.get_oldest_pending_lsn()` comment here, too.

File: pageserver/src/layered_repository/layer_map.rs (hunk 6 above)
Comment (patins) [reply pageserver/src/layered_repository/layer_map.rs]: Yeah, I get what you're saying. If there's a way to be more explicit while leaving out the other bound at min or max I think that'd be ok too

**Pull request merged**